src/sentry/models/apitoken.py

```diff
@@ -28,10 +29,98 @@ def default_expiration():
     return timezone.now() + DEFAULT_EXPIRATION
 
 
-def generate_token():
+def generate_token(token_type: AuthTokenType | str | None = AuthTokenType.__empty__) -> str:
+    if token_type:
+        return f"{token_type}{secrets.token_hex(nbytes=32)}"
+
     return secrets.token_hex(nbytes=32)
 
 
+class PlaintextSecretAlreadyRead(Exception):
+    def __init__(
+        self,
+        message="the secret you are trying to read is read-once and cannot be accessed directly again",
+    ):
+        super().__init__(message)
+
+
+class NotSupported(Exception):
+    def __init__(self, message="the method you called is not supported by this token type"):
+        super().__init__(message)
+
+
+class ApiTokenManager(ControlOutboxProducingManager):
+    def create(self, *args, **kwargs):
+        token_type: AuthTokenType | None = kwargs.get("token_type", None)
+
+        # Typically the .create() method is called with `refresh_token=None` as an
+        # argument when we specifically do not want a refresh_token.
+        #
+        # But if it is not None or not specified, we should generate a token since
+        # that is the expected behavior... the refresh_token field on ApiToken has
+        # a default of generate_token()
+        #
+        # TODO(mdtro): All of these if/else statements will be cleaned up at a later time
+        # to use a match statment on the AuthTokenType. Move each of the various token type
+        # create calls one at a time.
+        if "refresh_token" in kwargs:
+            plaintext_refresh_token = kwargs["refresh_token"]
+        else:
+            plaintext_refresh_token = generate_token()
+
+        if token_type == AuthTokenType.USER:
+            plaintext_token = generate_token(token_type=AuthTokenType.USER)
+            plaintext_refresh_token = None  # user auth tokens do not have refresh tokens
+        else:
+            # to maintain compatibility with current
+            # code that currently calls create with token= specified
+            if "token" in kwargs:
+                plaintext_token = kwargs["token"]
+            else:
+                plaintext_token = generate_token()
+
+        if options.get("apitoken.save-hash-on-create"):
+            kwargs["hashed_token"] = hashlib.sha256(plaintext_token.encode()).hexdigest()
+
+            if plaintext_refresh_token:
+                kwargs["hashed_refresh_token"] = hashlib.sha256(
+                    plaintext_refresh_token.encode()
+                ).hexdigest()
+
+        kwargs["token"] = plaintext_token
+        kwargs["refresh_token"] = plaintext_refresh_token
+
+        # Store the plaintext tokens for one-time retrieval
+        self.__plaintext_token = plaintext_token
+        self.__plaintext_refresh_token = plaintext_refresh_token
```

Comment:
This class is supposed to manage logic for all objects, I think storing the values in this class probably doesn't make sense? I could be wrong, and maybe this class gets instantiated per object and maybe this is totally safe, but I think the manager class shouldn't be used like this. I believe it gets instantiated once as part of the model, because it's a class attribute and not an instance attribute

Reply:
Based on this, i think the later property methods might have to change, so not going to comment on any of them yet

Reply:
![Screenshot 2024-04-10 at 11 49 22 AM](https://github.com/getsentry/sentry/assets/5581484/9e153c2e-3bc7-40ac-8670-206bfb92d84d)
Yeah we probably don't want this, does this example make sense?

Reply:
I think you want to add the values on the model object created, so
```
object_created = super().create(*args, **kwargs)
# somehow set these values but only once
object_created.__plaintext_token = plaintext_token
object_created.__plaintext_refresh_token = plaintest_refresh_token
```

Reply:
Very good catch! 🥔

Thank you for pairing with me on this the other day. I'm marking this as resolved since we've moved the logic to the actual model class to avoid this issue.